model-optimizer/extensions/ops/non_max_suppression.py

```diff
@@ -57,31 +56,53 @@ def backend_attrs(self):
 
     @staticmethod
     def infer(node: Node):
+        num_of_inputs = len(node.in_ports())
+        assert 2 <= num_of_inputs <= 6, \
+            "NonMaxSuppression node {} must have from 2 to 6 inputs".format(node.soft_get(node.name, node.id))
+
         boxes_shape = node.in_port(0).data.get_shape()
         assert boxes_shape is not None, 'The shape of tensor with boxes is not defined'
         scores_shape = node.in_port(1).data.get_shape()
         assert scores_shape is not None, 'The shape of tensor with scores is not defined'
         assert len(boxes_shape) == 3, 'Length of tensors with boxes must be equal to 3'
         assert len(scores_shape) == 3, 'Length of tensors with scores must be equal to 3'
 
-        max_output_boxes_per_class = node.in_port(2).data.get_value()
-        if max_output_boxes_per_class is None:
+        if num_of_inputs >= 3:
+            max_output_boxes_per_class = node.in_port(2).data.get_value()
+        else:
+            max_output_boxes_per_class = 0
+
+        if not max_output_boxes_per_class:
             log.info('Set default "max_output_boxes_per_class" for node {} to number of boxes'.format(node.name))
             max_output_boxes_per_class = boxes_shape[1]
 
         num_classes = scores_shape[1]
         num_input_boxes = boxes_shape[1]
         assert scores_shape[2] == num_input_boxes, 'Number of boxes mismatch'
 
-        if node.get_opset() == 'opset4':
+        if node.get_opset() in ['opset4', 'opset5']:
             max_number_of_boxes = min(num_input_boxes, max_output_boxes_per_class) * boxes_shape[0] * num_classes
         else:
             max_number_of_boxes = min(num_input_boxes, boxes_shape[0] * max_output_boxes_per_class * num_classes)
         node.out_port(0).data.set_shape(int64_array([max_number_of_boxes, 3]))
 
+        num_of_outputs = len([port for port in node.out_ports().values() if not port.disconnected()])
+        if num_of_outputs >= 2:
+            node.out_port(1).data.set_shape(int64_array([max_number_of_boxes, 3]))
+        if num_of_outputs >= 3:
+            node.out_port(2).data.set_shape(np.array(1).astype(np.int64))
+
     @staticmethod
     def type_infer(node):
-        if node.get_opset() in ['opset3', 'opset4']:
+        opset = node.get_opset()
+        if opset == 'opset5':
+            node.out_port(0).set_data_type(node.output_type)
+            num_of_outputs = len(node.out_ports())
+            if num_of_outputs >= 2:
```

Comment:
Check that output port #1 is connected instead. 

Similar comment for the line 103.

Reply:
If we write 
```python3
            if not node.out_port(1).disconnected(): 
                node.out_port(1).set_data_type(np.float32)
            if not node.out_port(2).disconnected():
                node.out_port(2).set_data_type(np.int64)
```
instead of lines 104-107, the we can get
```sh
mo.utils.error.Error: The data type for out port 2 of node ... is not defined
```
for a node with `type = `NonMaxSuppression` .

The reason of  that is lines 271-272
```python3
            log.debug('Type infer for node {}: {}'.format(node_name,
                                                          [port.get_data_type() for port in node.out_ports().values()]))
```
in the file `model-optimizer\mo\middle\passes\infer.py`.

Reply:
You may have 3 output ports but only 2 of them connected: 0 and 2 (1 is not connected). Your check for `num_of_outputs` will be incorrect.

You must check the the output port existence using `has_port` method and only if it exist then update the data type.